### PR TITLE
Core - add test vectors to test cases

### LINKED ISSUE #248: Core - Ed25519 signature test vectors
## Description

Add ed25519 signature [test vectors](https://github.com/iotaledger/tips/blob/main/tips/TIP-0014/tip-0014.md#test-vectors) into test case.

### PR DESCRIPTION
**Pull request by rebinmannady93**
# Description of change

Fixes #248 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

test_address.c

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests using CTest that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/core/utils/macros.h [hunk 1]
```diff
@@ -10,5 +10,7 @@
 #define BIN_TO_HEX_STR_BYTES(x) (BIN_TO_HEX_BYTES(x) + 1)
 // Get the string bytes of the given binary
 #define BIN_TO_STR_BYTES(x) (x + 1)
+// Get the binary bytes of the given hex string
```
Comment (matjazv) [inline src/core/utils/macros.h]: `hex string` or `hex bytes`?

File: src/crypto/iota_crypto.c [hunk 2]
```diff
@@ -120,6 +120,17 @@ int iota_crypto_sign(uint8_t const priv_key[], uint8_t msg[], size_t msg_len, ui
 #endif
 }
 
+int iota_crypto_sign_open(uint8_t msg[], size_t msg_len, uint8_t const pub_key[], uint8_t signature[]) {
+#if defined(CRYPTO_USE_SODIUM)
+  unsigned long long sign_len = ED_SIGNATURE_BYTES;
+  return crypto_sign_ed25519_open(msg, (unsigned long long *)&msg_len, signature, sign_len, pub_key);
```
Comment (matjazv) [inline src/crypto/iota_crypto.c]: `sign_len` variable is redundant.

```suggestion
  return crypto_sign_ed25519_open(msg, (unsigned long long *)&msg_len, signature, (unsigned long long)ED_SIGNATURE_BYTES, pub_key);
```

File: tests/crypto/test_vectors.h [hunk 3]
```diff
@@ -0,0 +1,102 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef __TEST_VECTORS_H__
+#define __TEST_VECTORS_H__
+
+#define TEST_VECTORS_COUNT 12
```
Comment (matjazv) [inline tests/crypto/test_vectors.h]: Because `test_vectors` are static, this is known at compile time.
```suggestion
#define TEST_VECTORS_COUNT (sizeof(test_vectors) / sizeof(test_vectors_t))
```

**Review by matjazv — CHANGES_REQUESTED** [review]

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -14,7 +14,7 @@ function(add_iota_test test_src test_name)
     target_link_libraries(${test_name} PRIVATE iota_core unity)
   elseif(${test_src} MATCHES "^crypto")
     add_dependencies(${test_name} iota_crypto)
-    target_link_libraries(${test_name} PRIVATE iota_crypto unity)
+    target_link_libraries(${test_name} PRIVATE iota_crypto iota_core unity)
```
Comment (oopsmonk) [inline tests/CMakeLists.txt]: crypto provides service for the core layer, therefore iota_crypto unit tests should be standalone, not consume iota_core.
https://iota-c-client.readthedocs.io/en/latest/client_intro.html#c-client-library-diagram

File: tests/crypto/test_crypto.c [hunk 5]
```diff
@@ -7,7 +7,10 @@
 #include <unistd.h>
 
 #include "blake2b_data.h"
+#include "core/utils/byte_buffer.h"
+#include "core/utils/macros.h"
```
Comment (oopsmonk) [inline tests/crypto/test_crypto.c]: remove unused headers. 
`HEX_TO_BIN_BYTES` is not wildly used, it can be defined here, not in `core/utils/macros.h`
```suggestion

```

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: tests/crypto/test_vectors.h (hunk 3 above)
Comment (rebinmannady93) [reply tests/crypto/test_vectors.h]: Thanks

File: src/core/utils/macros.h (hunk 1 above)
Comment (rebinmannady93) [reply src/core/utils/macros.h]: Hex bytes :-)

File: tests/CMakeLists.txt (hunk 4 above)
Comment (rebinmannady93) [reply tests/CMakeLists.txt]: I was using iota_core mainly for the hex_2_bin method in bute_buffer.c. I removed iota_core dependency and added hex_2_bin as static function in test_crypto.c

File: tests/crypto/test_crypto.c [hunk 6]
```diff
@@ -246,6 +282,36 @@ void test_pbkdf2_hmac_sha512() {
   }
 }
 
+void test_vector_signature_validity() {
```
Comment (oopsmonk) [inline tests/crypto/test_crypto.c]: the name doesn't clear to me, please express the source of vectors
```suggestion
// test vectors: https://github.com/novifinancial/ed25519-speccheck
void test_ed25519_signature_edge case() {
```

File: tests/crypto/test_crypto.c [hunk 7]
```diff
@@ -8,14 +8,50 @@
 
 #include "blake2b_data.h"
 #include "crypto/iota_crypto.h"
+#include "test_vectors.h"
```
Comment (oopsmonk) [inline tests/crypto/test_crypto.c]: please use a more specific file name, like 
```suggestion
#include "ed25519_signature_edges.h"
```

**Review by matjazv — APPROVED** [review]

File: tests/crypto/test_vectors.h [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+ * @brief test vector struct
+ *
+ */
+typedef struct {
+  char *address;
+  char *message;
+  char *pub_key;
+  char *signature;
+  bool valid;
+} test_vectors_t;
```
Comment (oopsmonk) [inline tests/crypto/test_vectors.h]:
```suggestion
} ed25519_edge_signature_t;
```

File: tests/crypto/test_vectors.h [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+static test_vectors_t test_vectors[] = {{"008f5a6fdcfef8989fb88312cbc956ccc54dceb9c693ec99c08a28bdda5f11da44",
```
Comment (oopsmonk) [inline tests/crypto/test_vectors.h]:
```suggestion
static ed25519_edge_signature_t ed25519_edge_sig[] = {{"008f5a6fdcfef8989fb88312cbc956ccc54dceb9c693ec99c08a28bdda5f11da44",
```

**Review by oopsmonk — CHANGES_REQUESTED** [review]
the implementation looks good to me. cleanup and renaming are needed to improve readability.

**Review by oopsmonk — APPROVED** [review]
